digest test : make more robust

edsantiago opened the pull request:
Instead of relying on random images from various separate
repositories, some of which are actually not the schema
version we thought we were testing (and some of which are
on unreliable registries, which causes test flakes),
craft our own images and store them all on quay.io

This commit:

  1) Adds a new script (tests/digest/make-v2sN) that
     crafts a custom image with the desired attributes:
     schema version 1 or 2, possibly with a duplicate
     layer. Script decides based on its name, that is,
     we are expected to have symlinks pointing to it.

  2) Updates digest.bats to use these newly-made images.

  3) Extends digest.bats by adding a check for 'v2s*'
     in the image name; if found, we now use 'imgtype'
     to extract the manifest and confirm that the
     pulled image has the expected schema version (1 or 2).

  4) Extends digest.bats by adding a check for duplicate
     layers in the one test that requires them.

Pushing the generated images to quay.io is left as an
exercise for the reader; as is updating the Digest SHAs
and/or tags in the individual tests in digest.bats

Fixes: #2147

Signed-off-by: Ed Santiago <santiago@redhat.com>

TomSweeneyRedHat commented:
I dare say you've outdone yourself this time @edsantiago 
It LGTM, but I'd like a head nod from @nalind about the digest testing as that's his baileywick.

edsantiago commented:
@nalind could I ask for your opinion and perhaps blessing on this?

nalind commented on tests/digest.bats:
```diff
@@ -3,40 +3,60 @@
 
 @test "from-by-digest-s1" {
-  fromreftest k8s.gcr.io/pause@sha256:bbeaef1d40778579b7b86543fe03e1ec041428a50d21f7a7b25630e357ec9247
+  fromreftest quay.io/libpod/testdigest_v2s1@sha256:816563225d7baae4782653efc9410579341754fe32cbe20f7600b39fc37d8ec7
```
Both of these are schema 1 images, so this is good.

nalind commented on tests/digest.bats:
```diff
@@ -3,40 +3,60 @@
 
 @test "from-by-tag-s1" {
-  fromreftest k8s.gcr.io/pause:0.8.0
+  fromreftest quay.io/libpod/testdigest_v2s1:20200210
```
Both of these are schema 1 images, so this is good.

nalind commented on tests/digest.bats:
```diff
@@ -3,40 +3,60 @@
 
 @test "from-by-digest-s2" {
-  fromreftest alpine@sha256:e9cec9aec697d8b9d450edd32860ecd363f2f3174c8338beb5f809422d182c63
+  fromreftest quay.io/libpod/testdigest_v2s2@sha256:755f4d90b3716e2bf57060d249e2cd61c9ac089b1233465c5c2cb2d7ee550fdb
```
This _was_ a schema 1 image, which was a bug in the test, and it's being replaced with a schema 2 image, which is a marked improvement.  Can't say enough good things about this.

nalind commented on tests/digest.bats:
```diff
@@ -3,40 +3,60 @@
 
 @test "from-by-tag-s2" {
-  fromreftest alpine:2.6
+  fromreftest quay.io/libpod/testdigest_v2s2:20200210
```
This was also a schema 1 image, which was a bug in the test, and it's being replaced with a schema 2 image, which again is a marked improvement.

nalind commented on tests/digest.bats:
```diff
@@ -3,40 +3,60 @@
 
 @test "from-by-repo-only-s2" {
-  fromreftest alpine
+  fromreftest quay.io/libpod/testdigest_v2s2
```
This became a manifest list, possibly after the test was added, which is a bug in the test.  This change fixes it, so again, very much an improvement.

nalind commented on tests/digest.bats:
```diff
@@ -3,40 +3,60 @@
+  dups=$(imgtype -expected-manifest-type '*' -show-manifest ${IMG} | jq .fsLayers|grep blobSum|sort|uniq -cd)
+  if [[ -z "$dups" ]]; then
+      die "Image ${IMG} does not have any duplicate layers (expected: one dup)"
+  fi
```
This test broke when we moved to using `libpod/whalesay`, since the new schema 1 image didn't have `fsLayers` and `history` lists in which with both the first and second entries were the identical.

I think we want to do a `jq '.fsLayers[].blobSum' | head -n 2 | uniq -cd` and `jq '.history[].v1Compatibility' | head -n 2 | uniq -cd` to detect that the image looks right, and find or make an image that looks more like `docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b` to use in the test.

edsantiago replied:
I'm super confused. Big wall o' questions coming.

First: when you say "this test broke", do you mean "the test started failing", or "the test was no longer testing what it was supposed to test"? I'm going to assume the latter, and that's exactly why my PR includes image tests that try to confirm image properties. (If you meant the former, or if I've misunderstood, please help me out).

Second... from my reading of git history, it looks like 86fa0803 introduced the `whalesay` image check and the next commit, ae7d2f35 less than one second later, also had something to do with duplicates. So I was expecting buildah-push to fail when run against 86fa0803^ -- and it does. But it fails in exactly the same way with ae7d2f35!
```console
++ ./buildah from --quiet --pull docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b
+ cid=whalesay-working-container-5
+ ./buildah push docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b dir:/tmp/bp/1
Getting image source signatures
Copying blob sha256:1154ba695078d29ea6c4e1adb55c463959cd77509adf09710e2315827d66271a
 188.05 MiB / 188.05 MiB [==================================================] 2s
Copying blob sha256:528c8710fd95f61d40b8bb8a549fa8dfa737d9b9c7c7b2ae55f745c972dddacd
ERRO[0002] error unmounting /var/lib/containers/storage/overlay/47a246479ce6a91248b69c2f79f3daabf527b3a2e4b68e4c0a1da8f8122d307f/merged: no such file or directory
error pushing image "docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b" to "dir:/tmp/bp/1": error copying layers and metadata: Error reading blob sha256:528c8710fd95f61d40b8bb8a549fa8dfa737d9b9c7c7b2ae55f745c972dddacd: error creating overlay mount to /var/lib/containers/storage/overlay/47a246479ce6a91248b69c2f79f3daabf527b3a2e4b68e4c0a1da8f8122d307f/merged: no such file or directory
```
What I am trying to do is figure out what is special about this whalesay, then get a version of buildah that fails with it, then get a version that passes, then try to craft my own custom image and see if we can use that instead of whalesay. Can you help me understand the buildah failures above, and help me isolate working/nonworking buildah commits for me to try out? Thank you!

edsantiago replied:
OBTW I think we should keep my `quay.io/libpod/testdigest_v2s1_with_dups` because with buildah ae7d2f35 it causes an interesting new failure:
```console
#  ./buildah commit testdigest_v2s1_with_dups-working-container-7 new-image
error committing container "testdigest_v2s1_with_dups-working-container-7" to "new-image": error copying layer and metadata: error reading configuration to write to image "containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]docker.io/library/new-image:latest": internal error: history lists 1 non-empty layers, but we have 4 layers on disk
```
So the exercise now is to create yet another magic image that causes old-buildah to fail but new-buildah to pass.

edsantiago replied:
DISCREPANCY SOLVED! I was using the default `overlay`, which of course doesn't work with buildah. Switching to `vfs` yields the expected results.

To recap: 86fa080 and ae7d2f35 are a two-part commit that addressed the whalesay image problem: after ae7d2f35, the `digest.bats` whalesay test passes. _However_, even with vfs and even with this commit, my hand-crafted `v2s1_with_dups` image causes buildah to fail:
```console
# ./buildah --storage-driver vfs commit testdigest_v2s1_with_dups-working-container-3 new-image
error committing container "testdigest_v2s1_with_dups-working-container-3" to "new-image": error copying layer and metadata: error reading configuration to write to image "containers-storage:[vfs@/var/lib/containers/storage+/var/run/containers/storage:vfs.override_kernel_check=true]docker.io/library/new-image:latest": internal error: history lists 1 non-empty layers, but we have 4 layers on disk
```
Given all this, I humbly suggest that my `testdigest_v2s1_with_dups` image is a better test than whalesay. Can we get this merged, and add new test cases (and images) as we find more failure cases?

nalind replied:
> I'm super confused. Big wall o' questions coming.
> 
> First: when you say "this test broke", do you mean "the test started failing", or "the test was no longer testing what it was supposed to test"? I'm going to assume the latter, and that's exactly why my PR includes image tests that try to confirm image properties. (If you meant the former, or if I've misunderstood, please help me out).

The latter, as the new image being used no longer had the characteristic that triggered the code paths we cared about for the test.

> Second... from my reading of git history, it looks like [86fa080](https://github.com/containers/buildah/commit/86fa0803e8991b2c970c1c248d8ff350c39a50ac) introduced the `whalesay` image check and the next commit, [ae7d2f3](https://github.com/containers/buildah/commit/ae7d2f354732f40e113163e8b2a214601517bae2) less than one second later, also had something to do with duplicates. So I was expecting buildah-push to fail when run against [86fa080](https://github.com/containers/buildah/commit/86fa0803e8991b2c970c1c248d8ff350c39a50ac)^ -- and it does. But it fails in exactly the same way with [ae7d2f3](https://github.com/containers/buildah/commit/ae7d2f354732f40e113163e8b2a214601517bae2)!
> 
> ```
> ++ ./buildah from --quiet --pull docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b
> + cid=whalesay-working-container-5
> + ./buildah push docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b dir:/tmp/bp/1
> Getting image source signatures
> Copying blob sha256:1154ba695078d29ea6c4e1adb55c463959cd77509adf09710e2315827d66271a
>  188.05 MiB / 188.05 MiB [==================================================] 2s
> Copying blob sha256:528c8710fd95f61d40b8bb8a549fa8dfa737d9b9c7c7b2ae55f745c972dddacd
> ERRO[0002] error unmounting /var/lib/containers/storage/overlay/47a246479ce6a91248b69c2f79f3daabf527b3a2e4b68e4c0a1da8f8122d307f/merged: no such file or directory
> error pushing image "docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b" to "dir:/tmp/bp/1": error copying layers and metadata: Error reading blob sha256:528c8710fd95f61d40b8bb8a549fa8dfa737d9b9c7c7b2ae55f745c972dddacd: error creating overlay mount to /var/lib/containers/storage/overlay/47a246479ce6a91248b69c2f79f3daabf527b3a2e4b68e4c0a1da8f8122d307f/merged: no such file or directory
> ```
> 
> What I am trying to do is figure out what is special about this whalesay, then get a version of buildah that fails with it, then get a version that passes, then try to craft my own custom image and see if we can use that instead of whalesay. Can you help me understand the buildah failures above, and help me isolate working/nonworking buildah commits for me to try out? 

The manifest in the image is a v2s1 manifest, and in that manifest, the first entries in both the `fsLayers` and `history` lists are identical to the second entries in those lists, which can be confirmed by using `skopeo inspect --raw`.  The duplicate values are in consecutive entries.

While `quay.io/libpod/testdigest_v2s1_with_dups@sha256:70d6c767101c907aa251c21fded459c0f1a481685eb764ca2f7a6162e24dd81a` looks like it would make a useful test, particularly if we fail to commit a new image based on it (or used to fail, but don't any more), it doesn't trigger duplicate elimination logic in containers/image that logic in buildah needs to agree with.

This can be confirmed by counting the number of `fsLayer` entries in the v2s1 manifest using `skopeo inspect --raw`, and comparing that number to the number of `rootfs.diff_ids` values we get when we ask `skopeo inspect --config` to produce a v2s2 configuration blob, which it has to generate from the manifest.  Part of the process for generating that configuration blob includes dropping the duplicates from the manifest.

`skopeo` doesn't have a CLI option for requesting the manifest in v2s2 format, but then this is the first occasion I can think of when one might come in handy.

edsantiago replied:
Thanks for your response. If I understand you correctly, the magic characteristics that are desired are that `.fsLayers` be greater than `.rootfs.diff_ids`, and any image with those characteristics should make a test case. Is that correct? 

If so, is there a trick to getting such an image uploaded to quay.io? I've hand-crafted an image (by editing `manifest.json`) such that I believe it matches those constraints:
```console
# skopeo inspect --raw containers-storage:7aaf3 | jq '.fsLayers | length'
WARN[0000] Not using native diff for overlay, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled
5

# skopeo inspect --config containers-storage:7aaf | jq '.rootfs.diff_ids | length'
WARN[0000] Not using native diff for overlay, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled
4
```
...but when I push it to quay.io, and pull it back, `.fsLayers` is back to four. I've tried the following push commands:
```console
# skopeo copy containers-storage:localhost/testdigest_v2s1_with_dups:latest docker://quay.io/libpod/testdigest_v2s1_with_dups:$(date +%Y%m%d)

# podman push localhost/testdigest_v2s1_with_dups:latest quay.io/libpod/testdigest_v2s1_with_dups:$(date +%Y%m%d)
```
And I always pull with `buildah`.

Is this something `quay.io` is doing, or is it my mistake in podman/skopeo push? Is there some way to get this image uploaded to quay, or is this a futile exercise?

nalind replied:
> Thanks for your response. If I understand you correctly, the magic characteristics that are desired are that `.fsLayers` be greater than `.rootfs.diff_ids`, and any image with those characteristics should make a test case. Is that correct?

No, their lengths relative to each other are not a key factor.  It's that the first two elements of the `fsLayers` are identical to each other, and the first two elements of the `history` are identical to each other.  The `history` values even have the same `id` and `parent` values in their `v1Compatibility` fields.

> Is this something `quay.io` is doing, or is it my mistake in podman/skopeo push? Is there some way to get this image uploaded to quay, or is this a futile exercise?

It appears to be something that happens at the client before it attempts to upload the manifest, normally a good thing.

edsantiago replied:
Thanks, Nalin. Using a `skopeo` doctored with your gist, I've created and uploaded `quay.io/libpod/testdigest_v2s1_with_dups:20200323`; and I've confirmed that upon pull, 'buildah manifest' includes the duplicate layers. 

Could I ask you to pull it and confirm that it meets your requirements for a test image? If it does, I'll spend the rest of the time updating the test scripts and documentation then resubmit for review.

nalind replied:
Yup, looks good!  Inspecting with `skopeo inspect --raw` shows the duplicates, and they're omitted from the configuration it synthesizes when I use `skopeo inspect --config`.

rhatdan replied:
Does this mean we can merge?

edsantiago replied:
Not right now. I made the stupid, stupid, stupid mistake of rebasing against master. Tests are broken, and it's nothing to do with anything in this PR. Right now it looks like #2245 but I can't understand how that passed CI. I'll keep investigating.

rhatdan commented:
@nalind PTAL so we can get this merged.

edsantiago commented:
Buildah gating tests are [failing on f32](https://jenkins-continuous-infra.apps.ci.centos.org/blue/organizations/jenkins/fedora-f32-build-pipeline/detail/fedora-f32-build-pipeline/63/pipeline/#step-248-log-1061), in digest tests. I saw similar failures on (manual) f31 tests yesterday.

Could we pretty-please get this reviewed and merged so we can all waste less time reviewing test failures?

edsantiago commented:
Well, I've given up. I can't figure out a way to create an equivalent container and preserve the characteristics when pushed to `quay.io`; I'm assuming that the bad whalesay image is a product of an earlier time, and all tools have checks now to make it impossible to push such an image to a modern repo.

I've removed the now-unused 'make-dup' functionality, but am leaving the `make-v2sN` script because it's still used for two other images and I believe it's important to have reproducible test images. I've undone the `quay.io/whalesay` edit, reverting back to `docker.io/whalesay`, and have added a minimal sanity check to make sure we don't accidentally go back to a different image. I've also added two retries in hopes of alleviating the common whalesay flakes.

I'd still like to get this reviewed and merged, because I still believe it's a more accurate and robust test.

edsantiago commented:
Gating tests [are failing](https://jenkins-continuous-infra.apps.ci.centos.org/blue/organizations/jenkins/fedora-rawhide-build-pipeline/detail/fedora-rawhide-build-pipeline/9993/pipeline/#step-248-log-1060); this PR would fix that.

TomSweeneyRedHat commented on tests/digest.bats:
```diff
@@ -3,40 +3,92 @@
+  # this image was created, and there no longer seem to be any tools that
+  # can create a similar image. (Yes, this means our testing is at the
+  # mercy of docker.io not removing that precise image from their registry)
+  IMG=docker.io/docker/whalesay@sha256:178598e51a26abbc958b8a2e48825c90bc22e641de3d31e18aaf55f3258ba93b
```
should we snab a copy of the image and tuck it away in quay.io somwehere?

TomSweeneyRedHat replied:
regardless of the answer,  I wouldn't hold the merge back for this question.

edsantiago replied:
We can't: that's the entire point of my wasted few weeks of effort.

It seems like this image was magically written by earlier tools that (waves hands) did something that modern tools can no longer do. In other words: it is impossible to push this image anywhere and have it keep the properties that make it interesting. Current tools (buildah, skopeo, podman) somehow flatten/smoosh/deduplicate the layers on any new write. All we have is this image, on docker.io, and all we can do is hope it never changes.

Disclaimer: I have no idea what the technical terms are for any of this; I just know I've tried really really hard for several weeks and all my attempts fail on push.

nalind replied:
The current versions of our tools are dutifully cleaning up confusing bits of old data, which is rather inconvenient for us when we want to preserve those bits.

After some experimentation, it looks like having skopeo's vendored copy of containers/image not call `s1.fixManifestLayers()` in `github.com/containers/image/v5/manifest.Schema1FromManifest()` is enough to let it copy the image from docker hub to my test repository with its duplicate `blobSum` and `v1Compatibility` values intact.

I uploaded the change I made relative to [skopeo master](https://github.com/containers/skopeo/tree/e31d5a0e8fc61241b02b6eb06aeeb39ed4de7b5e) as a gist [here](https://gist.github.com/nalind/b491204ff05c3c3f3b6ef014b333a60c).  Hopefully it'll help preserve this detail for us.

TomSweeneyRedHat commented:
LGTM and happy green test buttons.  Would like a head nod from @nalind if possible before merging.

edsantiago commented:
Can someone with privs rerun the failing test? I think it's a flake.

The more I look at it, the less I understand it: the test that failed is `something-http-pre` and is bytewise identical to the test immediately after, which is `something-http-post`... which passes.  There are two errors:
```
[+0312s] # 2020/03/25 20:14:00 listen tcp :33192: bind: address already in use
[+0312s] # error prepping temporary context directory: error getting "http://0.0.0.0:33192/context.tar": Get http://0.0.0.0:33192/context.tar: dial tcp 0.0.0.0:33192: connect: connection refused
```
I'm guessing the first one is from `httpd`, the second from `buildah` (the helper script starts httpd but doesn't actually confirm that it succeeds). I don't know why port 33192 would be in use.

Finally, I can't reproduce it on my own test setup. Has anyone seen this before?

edsantiago commented:
Yay, tests green!

For ease of review, I've left this as two commits. The second is the one that adds and documents the dup-layer image.

rhatdan commented:
bors r+

edsantiago commented:
Followup: buildah gating tests [are green](https://bodhi.fedoraproject.org/updates/FEDORA-2020-7f2d2a418f); yesterday's errors (timeouts on docker.io) probably deserve investigation and fixing so we continue removing our reliance on docker.io, but I can get to that next week.

Thanks again @nalind for your patience and help with this one.
